### ISSUE TITLE
Node status is unhealthy after creating the DC/OS cluster.

### ISSUE THREAD

**Issue by wangtt03** [Azure/acs-engine#565]
I use the latest acs-engine from GitHub and deploy a DC/OS cluster (default should be 1.9.0) on global azure and always have 1 or 2 unhealthy node. The dcos-navstar.service is not started correctly, the log is as following:
```
dcos-navstar.service - Navstar: A distributed systems & network overlay orchestration engine
   Loaded: loaded (/opt/mesosphere/packages/navstar--1128db0234105a64fb4be52f4453cd6aa895ff30/dcos.target.wants/dcos-navstar.service; enabled; vendor preset: enabled)
   Active: activating (auto-restart) (Result: exit-code) since Sun 2017-04-16 06:41:40 UTC; 3s ago
  Process: 11540 ExecStartPre=/opt/mesosphere/bin/check-time (code=exited, status=1/FAILURE)
 
Apr 16 06:41:40 dcos-agentprivate-13963398000000 systemd[1]: Failed to start Navstar: A distributed systems & network overlay orchestration engine.
Apr 16 06:41:40 dcos-agentprivate-13963398000000 systemd[1]: dcos-navstar.service: Unit entered failed state.
Apr 16 06:41:40 dcos-agentprivate-13963398000000 systemd[1]: dcos-navstar.service: Failed with result 'exit-code'.
```
I tried to `systemctl restart system-timesyncd.service` then the dcos-navstar.service started. I did some deployments on version 1.7.3/1.8.7/1.8.7 and all these deployments have the same problems.
 
I also tried to deploy it on the AzureChina and the same error exists but a little bit worse than global (4-5 unhealthy agents plus master job service failure). Manually restart the system-timesyncd.sevice will solve the problem. Did you see this error before? Thanks!

**Comment by JackQuincy** [comment]
@rgardler Is this something you have heard of before?

**Comment by adragomir** [comment]
@wangtt03 @JackQuincy @rgardler this seems to be related to some issues in HyperV / Linux time synchronization
Some details can be found in these issues: 
1. https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1676635
2. https://patchwork.kernel.org/patch/9525945/

For us, the patch was to: 

1. Get this on all the nodes https://raw.githubusercontent.com/torvalds/linux/master/tools/hv/lsvmbus
2. Use the script to find out the VMBUS ID for the Time Synchronization service:

``` 
VMBUS ID 11: Class_ID = {9527e630-d0ae-497b-adce-e80ab0175caf}
- Time Synchronization
Device_ID =
{2dd1ce17-079e-403c-b352-a1921ee207ee}
Sysfs path: /sys/bus/vmbus/devices/2dd1ce17-079e-403c-b352-a1921ee207ee
Rel_ID=11, target_cpu=0
```
3. Pass this to hv_util unbind, to disable time syncing with the underlying host:  `echo 2dd1ce17-079e-403c-b352-a1921ee207ee | sudo tee /sys/bus/vmbus/drivers/hv_util/unbind`
4. Re-run `systemctl restart systemd-timesyncd` and `hwclock --systohc --utc` until time shows green. 
5. You can run the check by hand, by doing `ENABLE_CHECK_TIME=true check-time`

**Comment by MarcinMilewski** [comment]
+ 1

**Comment by firedog1024** [comment]
@MarcinMilewski @adragomir 

We are also seeing the exact same issue when we apply the fix provided by adragomir.  If anyone can provide a solution to get systemd-timesyncd up and running again it would be most appreciated

**Comment by MarcinMilewski** [comment]
Temporary fix (didn't test reboot yet. Probably need to make systemctl disable systemd-timesyncd and add ntp to start on boot ):
sudo apt-get install ntp
sudo echo 2dd1ce17-079e-403c-b352-a1921ee207ee | sudo tee /sys/bus/vmbus/drivers/hv_util/unbind
sudo systemctl stop systemd-timesyncd
sudo hwclock --systohc --utc
sudo ntpdate -s ntp.ubuntu.com
sudo service ntp stop
sudo ntpd -gq
sudo service ntp start
ENABLE_CHECK_TIME=true /opt/mesosphere/bin/check-time

Is it possible to set Ubuntu base vm to 14.04 from 16.04 ?

**Comment by rafarhat** [comment]
@MarcinMilewski We were able to get system-timesyncd up and running with @adragomir workaround by removingPrivateTmp in the system-timesyncd.service unit file. For some reason after a physical reboot we aren't able to get system-timesyncd running with a PrivateTmp=true.

**Comment by adragomir** [comment]
@MarcinMilewski @firedog1024 I think the differences are because I'm actually using CoreOS, not Ubuntu, and there is a high probability that's where the differences come from. Sorry, should have made that clear. Also, I should have made clear that we don't yet have a final automated solution, which is probably a bit different depending on the underlying OS.

**Comment by MarcinMilewski** [comment]
@adragomir Thank you for your help.
I can confirm that solution with different osImageVersion and osImageSKU for Ubuntu works.
After template generation need to modify azuredeploy.json and change _osImageSku_ and _osImageVersion_ for the following:
`    "osImageSKU": "16.04.0-LTS",
    "osImageVersion": "16.04.201606270",`

**Comment by MarcinMilewski** [comment]
From https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1676635:
This bug was fixed in the package linux - 4.8.0-54.57. Could you make possible to use new osImageVersion with these fixes ?

**Comment by jmspring** [comment]
@MarcinMilewski your recommendation helps with making sure everything runs on initial deployment.  Unfortunately, the Navstar issue comes back after reboot.

It would be good to have this bug fixed, I agree.

@JackQuincy any chance of getting this in ACS engine?

**Comment by JackQuincy** [comment]
@jmspring  I won't pretend to have followed this thread. Is the fix in question going to a newer version of Ubuntu? is so we are currently on
"osImageSKU": "16.04-LTS", 
    "osImageVersion": "16.04.201705160",
Which I would assume is good enough. Does that fix the issue or is the bug in this version as well?

**Comment by jmspring** [comment]
The bug is in this version as well.  Unfortuntately, the patch by Linux has
been applied to 16.10 and 17.04 it appears but is not yet in 16.04.  This
is a bit problematic since I'm working with a customer who is looking to
spin up an DC/OS cluster using acs-engine (mixed agent pools, etc).

**Comment by MarcinMilewski** [comment]
@jmspring did you try:
"osImageSKU": "16.04.0-LTS",
"osImageVersion": "16.04.201606270" ?

Using this setting I have no problems with Navstar issue after reboot.

**Comment by jmspring** [comment]
@MarcinMilewski - I was using an existing azuredeploy.json where I had that
change.  I'll try again.  But I'm writing some services that
install/configure some drivers and do a couple of reboots.  Testing those
is when I saw it.

On Jun 20, 2017 1:49 AM, "MarcinMilewski" <notifications@github.com> wrote:

@jmspring <https://github.com/jmspring> did you try:
"osImageSKU": "16.04.0-LTS",
"osImageVersion": "16.04.201606270" ?

Using this setting I have no problems with Navstar issue after reboot.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Azure/acs-engine/issues/565#issuecomment-309687596>,
or mute
the thread

**Comment by JackQuincy** [comment]
https://github.com/Azure/acs-engine/pull/817 should actually fix this issue

**Comment by MarcinMilewski** [comment]
Packages update on `16.04.201606270` makes problem with time synchronization again.
@JackQuincy 
#817 Didn't test with a new cluster, but run these lines (from the merge pull commit) on current Ubuntu 16.04 node doesn't fix it:
`echo 2dd1ce17-079e-403c-b352-a1921ee207ee > /sys/bus/vmbus/drivers/hv_util/unbind`
`sed -i "13i\echo 2dd1ce17-079e-403c-b352-a1921ee207ee > /sys/bus/vmbus/drivers/hv_util/unbind\n" /etc/rc.local`

Currently I I have managed to apply this workaround successfully. It disables _systemd-timesyncd_ service and enables _ntp_ at boot. I have no problem with _Navstar_ after reboot, but I do not know the possible consequences of  _systemd-timesyncd_ replacement unfortunately.
`sudo apt-get install ntp ntpstat
sudo echo 2dd1ce17-079e-403c-b352-a1921ee207ee | sudo tee /sys/bus/vmbus/drivers/hv_util/unbind
sudo sed -i "13i\echo 2dd1ce17-079e-403c-b352-a1921ee207ee > /sys/bus/vmbus/drivers/hv_util/unbind\n" /etc/rc.local
sudo systemctl stop systemd-timesyncd
sudo service ntp stop
sudo ntpd -gq
sudo service ntp start
ENABLE_CHECK_TIME=true /opt/mesosphere/bin/check-time
sudo systemctl disable systemd-timesyncd
sudo systemctl enable ntp`

**Comment by julienlau** [comment]
+1 for a quickfix
very annoying

**Comment by shawnmcknight** [comment]
I'm echoing @MarcinMilewski a bit here, but I did spin up a new cluster today with #817 in place.  My `/etc/rc.local` file is reflecting that the change from that PR took effect:
```
#!/bin/sh -e
#
# rc.local
#
# This script is executed at the end of each multiuser runlevel.
# Make sure that the script will "exit 0" on success or any other
# value on error.
#
# In order to enable or disable this script just change the execution
# bits.
#
# By default this script does nothing.
echo 2dd1ce17-079e-403c-b352-a1921ee207ee > /sys/bus/vmbus/drivers/hv_util/unbind


exit 0
```
After rebooting a VM within the cluster, the Navstar service comes back unhealthy.  On the master node, the Metronome service comes back unhealthy.  I don't believe #817 resolved the issue it was hoping to.  I did also confirm that the steps @MarcinMilewski laid out to replace timesyncd with ntp does cause the services to come back healthy, but also do not know if there are any side effects to having done those steps.

cc @JackQuincy @anhowe

**Comment by shawnmcknight** [comment]
In reviewing https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1676635 I am unsure that is the issue that is actually causing the problem with starting the systemd-timesyncd.service upon reboot.  That issue doesn't mention anything along those lines and from what I can see it has been fixed in 16.04.  Updating the server put me beyond the kernel version mentioned in the 07-24 update (4.4.0-87.110) but the problem persisted.

Beyond that, I wanted to see if I could automate the approach that @MarcinMilewski laid out and see if that mitigated the issue.  I modified the CustomData for the `#cloud-config` for all nodes in a deployment to install the ntp and ntpstat packages and run the series of commands to disable the timesyncd service and enable the ntp service.  I then deployed that update and tested a reboot.

After rebooting, it took a while for `ntpstat` to report as synchronized and a little while longer for `ENABLE_CHECK_TIME=true /opt/mesosphere/bin/check-time` to report that time was in sync.  They both eventually did and at that point all services were coming back healthy.

I'll reiterate again that I also have no idea if this is a 'good' approach, but being unable to reboot any server in the dc/os cluster without services going unhealthy doesn't seem tenable.  Resolving the root cause of systemd-timesyncd not starting would be ideal, but there doesn't appear to be any movement on that so this might be a good alternative.  Thoughts?

**Comment by shawnmcknight** [comment]
Based on the diagnosis of this problem being as a result of the `systemd.timesyncd` service not starting, I thought I'd see how other Ubuntu instances outside of ACS behaved.  I was unable to replicate a problem on a simple Ubuntu VM created in Azure, so I decided that it was a problem within ACS.

TL;DR - I believe this problem is caused by exactly the same issue as #1099 and presumably fixed by #1100.  Read on for details.

`systemd-timesyncd.service` was not starting due to a nebulous error regarding a dependency.  I eventually found that this service was reporting the following dependencies:
```
sudo systemctl list-dependencies systemd-timesyncd
systemd-timesyncd.service
● ├─-.mount
● ├─system.slice
● ├─var-tmp.mount
● └─time-sync.target
```
The `var-tmp.mount` dependency was in a failed state AND this dependency did not exist in my non-ACS VM instance.  After reviewing the Azure deployment configuration, I found that there were 3 ephemeral filesystems and only 1 of them (`/var/lib/docker`) was mounting. `/var/lib/mesos` and `/var/tmp` were not mounting and I was unable to make the mount happen.  Essentially there was a cascading dependency tree which was preventing DC/OS services such as navstar from starting properly.  That is, `var-tmp.mount` failed which caused `systemd-timesyncd.service` to fail which caused `dcos-navstar.service` to fail.

This also explains why @rafarhat was able to resolve his problem by changing the `PrivateTmp` setting in the service definition for `systemd-timesyncd.service`.  By altering that setting, the dependency on `var-tmp.mount` was removed for `systemd-timesyncd` and that service was able to start which also allowed navstar to start.

As I see it, there are now several ways to work around this issue, although it would be ideal if #1100 could be expedited into production.  I have tested replacing the `systemd-timesyncd` service with ntp and that fixed the issue.  I have also tested removing the mounts for `/var/tmp` and `/var/lib/mesos` from `/etc/fstab` and that fixed the issue.  Although I haven't tested it, I have every reason to believe that removing the `PrivateTmp` flag from `/lib/systemd/system/systemd-timesyncd.service` will work as well.  I'm sure each of these carries its own potential issues/problems and none of them resolve the underlying issue of the filesystems not mounting.  However, anyone encountering this issue can pick the path of least resistance by employing one of those methods until #1100 has been merged.
